bokeh/model.py

```diff
@@ -59,30 +59,40 @@
 # Dev API
 #-----------------------------------------------------------------------------
 
-def collect_models(*input_values):
+def collect_filtered_models(discard, *input_values):
     ''' Collect a duplicate-free list of all other Bokeh models referred to by
-    this model, or by any of its references, etc.
+    this model, or by any of its references, etc, unless filtered-out by the
+    provided callable.
 
     Iterate over ``input_values`` and descend through their structure
-    collecting all nested ``Models`` on the go. The resulting list is
-    duplicate-free based on objects' identifiers.
+    collecting all nested ``Models`` on the go.
 
     Args:
+        discard (Callable[[Model], bool])
+            a callable returning whether to continue down the object's references
         *input_values (Model)
             Bokeh models to collect other models from
 
     Returns:
-        list[Model] : all models reachable from this one.
+        None
 
     '''
 
     ids = set([])
     collected = []
     queued = []
 
-    def queue_one(obj):
-        if obj.id not in ids:
-            queued.append(obj)
+    queue_one = None
+    if discard is None:
+        def _visitor(obj):
+            if obj.id not in ids:
+                queued.append(obj)
+        queue_one = _visitor
+    else:
+        def _filteredvistor(obj):
+            if obj.id not in ids and not discard(obj):
+                queued.append(obj)
+        queue_one = _filteredvistor
```

Comment:
@poldavezac how about simplifying this by using a default filter in all cases:
```
if discard is None: 
    discard = lambda x: False

def _visitor(obj):
    if obj.id not in ids and not discard(obj)::
        queued.append(obj)
    queue_one = _visitor
```
Given the relatively small number of total objects, I don't think there is a performance concern to justify specialized code paths in this case (but certainly open to evidence otherwise)